### PR TITLE
使用深度学习自动生成HTML代码 - 第 1 部分

### LINKED ISSUE #3089: 使用深度学习自动生成HTML代码 - 第 1 部分
* 原文链接：[Turning Design Mockups Into Code With Deep Learning - Part 1](https://blog.floydhub.com/turning-design-mockups-into-code-with-deep-learning/)
  * Markdown文件：[文件地址](https://github.com/xitu/gold-miner/blob/master/TODO/turning-design-mockups-into-code-with-deep-learning-1.md)
  * PR 地址：#3184
  * 文章分类：**机器学习**
  ----
  * 翻译时间：8 天
  * 校对时间：2 天
  * 翻译积分：8 分
  * 校对积分：2.5 分
  * [积分有什么用](https://github.com/xitu/gold-miner/wiki)

### PR DESCRIPTION
**Pull request by sakila1012**
译文翻译完成，resolve #3089

### CODE REVIEW

**Comment by sunshine940326** [comment]
校对认领

**Comment by fanyijihua** [comment]
@sunshine940326 好的呢 :beer:

**Comment by wzy816** [comment]
校对认领

**Comment by fanyijihua** [comment]
@wzy816 妥妥哒 :beers:

File: TODO/turning-design-mockups-into-code-with-deep-learning-1.md [hunk 1]
```diff
@@ -178,147 +177,149 @@ Here we make the prediction:
...
+然后，解码器将结合原型设计特征和标记特征以创建下一个标签的特征，这一个特征可以通过全连接层预测下一个标签。
 
-##### Design mockup features
+##### 设计原型的特征
 
-Since we need to insert one screenshot for each word, this becomes a bottleneck when training the network ([example](https://docs.google.com/spreadsheets/d/1xXwarcQZAHluorveZsACtXRdmNFbwGtN3WMNhcTdEyQ/edit#gid=0)). Instead of using the images, we extract the information we need to generate the markup.
+因为我们需要为每个单词插入一个截屏，这将会成为训练神经网络[案例](https://docs.google.com/spreadsheets/d/1xXwarcQZAHluorveZsACtXRdmNFbwGtN3WMNhcTdEyQ/edit#gid=0))的瓶颈。因此我们抽取生成标记语言所需要的信息来替代直接使用图像。
 
-The information is encoded into image features. This is done by using an already pre-trained convolutional neural network (CNN). The model is pre-trained on Imagenet.
+这些抽取的信息将通过预训练的 CNN 编码到图像特征中。这个模型是在图像上预训练的。
```
Comment (wzy816) [inline TODO/turning-design-mockups-into-code-with-deep-learning-1.md]: 这个模型是在图像上预训练的。->这个模型是在Imagenet上预先训练好的。

File: TODO/turning-design-mockups-into-code-with-deep-learning-1.md [hunk 2]
```diff
@@ -178,147 +177,149 @@ Here we make the prediction:
...
-We’ll take the word embeddings and run them through an LSTM and return a sequence of markup features. These are run through a Time distributed dense layer - think of it as a dense layer with multiple inputs and outputs.
+我们现在将词嵌入馈送到 LSTM 中，并期望能返回一系列的标记特征。这些标记特征随后会馈送到一个 Time Distributed 密集层，该层级可以视为有多个输入和输出的全连接层。
 
 ![](https://blog.floydhub.com/static/encoder-78498407f393e83128abed5eec86dd4c-6c1a3.png) 
 
-In parallel, the image features are first flattened. Regardless of how the digits where structured they are transformed into one large list of numbers. Then we apply a dense layer on this layer to form a high-level features. These image features are then concatenated to the markup features.
+对于另一个平行的过程，其中图像特征首先会展开成一个向量，然后再馈送到一个全连接层而抽取出高级特征。这些图像特征随后会与标记特征相级联而作为编码器的输出。
 
-This can be hard to wrap your mind around - so let’s break it down.
+这很难把理解 - 因此将它分解。
```
Comment (wzy816) [inline TODO/turning-design-mockups-into-code-with-deep-learning-1.md]: 这很难把理解 - 因此将它分解。 - >这个有点难理解，让我来分步描述一下。

File: TODO/turning-design-mockups-into-code-with-deep-learning-1.md [hunk 3]
```diff
@@ -178,147 +177,149 @@ Here we make the prediction:
...
 ![](https://blog.floydhub.com/static/word_embedding_markup_feature-d4e76483527fefd10742c0ddc1cd3227-6c1a3.png) 
 
-To mix signals and find higher-level patterns we apply a TimeDistributed dense layer to the markup features. TimeDistributed dense is the same as a dense layer but with multiple inputs and outputs.
+为了混合信号并寻找高级模式，我们运用了一个 TimeDistributed 密集层以抽取标记特征。TimeDistributed 密集层和一般的全连接层非常相似，且它有多个输入与输出。
 
-##### Image features
+##### 图像特征
 
-In parallel, we prepare the images. We take all the mini image features and transform them into one long list. The information is not changed, just reorganized.
+对于另一个平行的过程，我们需要将图像的所有像素值展开成一个向量，因此信息不会被改变，它们只会用来识别。
```
Comment (wzy816) [inline TODO/turning-design-mockups-into-code-with-deep-learning-1.md]: 对于另一个平行的过程 -> 同时

Comment (wzy816) [inline TODO/turning-design-mockups-into-code-with-deep-learning-1.md]: 它们只会用来识别 -> 只是重组了一下

File: TODO/turning-design-mockups-into-code-with-deep-learning-1.md [hunk 4]
```diff
@@ -178,147 +177,149 @@ Here we make the prediction:
...
-Note that the LSTM layer has the sequence set to false. Instead of returning the length of the input sequence it only predicts one feature. In our case, it’s a feature for the next tag. It contains the information for the final prediction.
+注意 LSTM 层不应该返回一个长度等于输入序列的向量，而只需要预测预测一个特征。在我们的案例中，这个特征将预测下一个标签，它包含了最后预测的信息。
 
 ![](https://blog.floydhub.com/static/image-markup-feature_to_vocab-eb39368b3f466914c9383d532675a622-6c1a3.png) 
 
-##### The final prediction
+##### 最后的预测
 
-The dense layer works like a traditional feedforward neural network. It connects the 512 digits in the next tag feature with the 4 final predictions. Say we have 4 words in our vocabulary: start, hello, world, and end.
+密集层会像传统前馈网络那样工作，它将下一个标签特征中的 512 个值与最后的四个预测连接起来，即我们在词汇表所拥有的四个单词：start、hello、world 和 end。
```
Comment (wzy816) [inline TODO/turning-design-mockups-into-code-with-deep-learning-1.md]: 密集层 -> 全连接层
之前都翻译得对呀

File: TODO/turning-design-mockups-into-code-with-deep-learning-1.md [hunk 5]
```diff
@@ -178,147 +177,149 @@ Here we make the prediction:
...
 
-The vocabulary prediction could be [0.1, 0.1, 0.1, 0.7]. The softmax activation in the dense layer distributes a probability from 0 - 1, with the sum of all predictions equal to 1\. In this case, it predicts that the 4th word is the next tag. Then you translate the one-hot encoding [0, 0, 0, 1] into the mapped value, say “end”.
+词汇的预测值可能是 [0.1, 0.1, 0.1, 0.7]。密集层最后采用的 softmax 函数会为四个类别产生一个 0-1 概率分布，所有预测值的和等于 1。在这个案例中，例如将预测第四个词为下一个标签。然后，你可以将 one-hot 编码 [0, 0, 0, 1] 转译成映射的值，也就是 “end”。
```
Comment (wzy816) [inline TODO/turning-design-mockups-into-code-with-deep-learning-1.md]: softmax 函数 -> softmax激活函数

File: TODO/turning-design-mockups-into-code-with-deep-learning-1.md [hunk 6]
```diff
@@ -6,91 +6,91 @@
...
+- [使用深度学习自动生成HTML代码 - 第 2 部分](https://github.com/xitu/gold-miner/blob/master/TODO/turning-design-mockups-into-code-with-deep-learning-2.md)
 
-Within three years deep learning will change front-end development. It will increase prototyping speed and lower the barrier for building software.
+在未来三年来，深度学习将改变前端的发展。它将会加快原型设计的速度和降低构建软件的门槛。
 
-The field took off last year when Tony Beltramelli introduced the [pix2code paper](https://arxiv.org/abs/1705.07962) and Airbnb launched [sketch2code](https://airbnb.design/sketching-interfaces/).
+Tony Beltramelli 去年发布了[pix2code 论文](https://arxiv.org/abs/1705.07962)，Airbnb 也发布了 [sketch2code](https://airbnb.design/sketching-interfaces/)。
 
-Currently, the largest barrier to automating front-end development is computing power. However, we can use current deep learning algorithms, along with synthesized training data, to start exploring artificial front-end automation right now.
+目前，自动化前端开发的最大屏障是计算能力。但我们可以使用目前的深度学习算法，以及同步训练数据来探索人工智能前端自动化的方法。
```
[the remark this answers is not in the document]
Comment (sakila1012) [reply TODO/turning-design-mockups-into-code-with-deep-learning-1.md]: 恩，谢谢

File: TODO/turning-design-mockups-into-code-with-deep-learning-1.md [hunk 7]
```diff
@@ -2,95 +2,95 @@
...
+## 核心逻辑
 
-Let’s recap our goal. We want to build a neural network that will generate HTML/CSS markup that corresponds to a screenshot.
+让我们回顾一下我们的目标。我们的目标是构建一个神经网络，能够生成与截图对应的 HTML/CSS。
 
-When you train the neural network, you give it several screenshots with matching HTML.
+当你训练神经网络时，你先提供几个截图和对应的 HTML 代码。
 
-It learns by predicting all the matching HTML markup tags one by one. When it predicts the next markup tag, it receives the screenshot as well as all the correct markup tags until that point.
+网络通过逐个
```
Comment (sunshine940326) [inline TODO/turning-design-mockups-into-code-with-deep-learning-1.md]: 这里是不是没有翻译完？

File: TODO/turning-design-mockups-into-code-with-deep-learning-1.md [hunk 8]
```diff
@@ -2,95 +2,95 @@
...
+这里是一个在 Google Sheet [简单的训练数据示例](https://docs.google.com/spreadsheets/d/1xXwarcQZAHluorveZsACtXRdmNFbwGtN3WMNhcTdEyQ/edit?usp=sharing)。
 
-Here is a simple [training data example](https://docs.google.com/spreadsheets/d/1xXwarcQZAHluorveZsACtXRdmNFbwGtN3WMNhcTdEyQ/edit?usp=sharing) in a Google Sheet.
+创建逐词预测的模型是现在最常用的方法。这里也有[其他方法](https://machinelearningmastery.com/deep-learning-caption-generation-models/)，但该方法也是本教程使用的方法。
```
Comment (sunshine940326) [inline TODO/turning-design-mockups-into-code-with-deep-learning-1.md]: 也是本教程使用的方法。 ->

**Comment by sakila1012** [comment]
@sunshine940326 已改

File: TODO/turning-design-mockups-into-code-with-deep-learning-1.md [hunk 9]
```diff
@@ -178,147 +177,149 @@ Here we make the prediction:
...
 floyd run --gpu --env tensorflow-1.4 --data emilwallner/datasets/imagetocode/2:data --mode jupyter
 ```
 
-All the notebooks are prepared inside the floydhub directory. The local equivalents are under local. Once it’s running, you can find the first notebook here: floydhub/Hello_world/hello_world.ipynb .
-
-If you want more detailed instructions and an explanation for the flags, check [my earlier post](https://blog.floydhub.com/colorizing-b&w-photos-with-neural-networks/).
+所有的 notebooks 都放在 floydbub 目录下。本地等同于本地目录下。一旦我们开始运行模型，那么在 floydhub/Helloworld/helloworld.ipynb 下可以找到第一个 Notebook。
 
-## HTML Version
+如果你想了解更多的指南和对 flags 的解释，请查看本项目[早期的文章](https://blog.floydhub.com/colorizing-b&w-photos-with-neural-networks/)。
```
Comment (sunshine940326) [inline TODO/turning-design-mockups-into-code-with-deep-learning-1.md]: 请查看本项目[早期的文章] -> 请查看我[早期的文章]

**Comment by sunshine940326** [comment]
我还没有校完，中午去吃个饭，译者改的太速度了，太给力了，稍等下还有几个问题，好了艾特你
@sakila1012

File: TODO/turning-design-mockups-into-code-with-deep-learning-1.md [hunk 10]
```diff
@@ -178,147 +177,149 @@ Here we make the prediction:
...
-### Overview
+### 概览
 
-If we expand the components of the previous graphic it looks like this.
+如果我们将前面的架构扩展为以下图展示的结构。
 
 ![](https://blog.floydhub.com/static/model_more_details-68db3bf26f6df205ffe4c541ace33a92-6c1a3.png) 
 
-There are two major sections. First, the encoder. This is where we create image features and previous markup features. Features are the building blocks that the network creates to connect the design mockups with the markup. At the end of the encoder, we glue the image features to each word in the previous markup.
+该架构主要有两个部分。首先，编码器。编码器是我们创建图像特征和前面标记特征（markup features）的部分。特征是网络创建原型设计和标记语言之间联系的构建块。在编码器的末尾，我们将图像特征传递给前面标记的每一个单词。
```
Comment (sunshine940326) [inline TODO/turning-design-mockups-into-code-with-deep-learning-1.md]: 首先，编码器。编码器是我们创建图像特征和前面标记特征（markup features）的部分 -> 首先，编码器。编码器是我们创建图像特征和前面标记特征（markup features）的地方

File: TODO/turning-design-mockups-into-code-with-deep-learning-1.md [hunk 11]
```diff
@@ -178,147 +177,149 @@ Here we make the prediction:
...
 
-The decoder then takes the combined design and markup feature and creates a next tag feature. This feature is run through a fully connected neural network to predict the next tag.
+然后，解码器将结合原型设计特征和标记特征以创建下一个标签的特征，这一个特征可以通过全连接层预测下一个标签。
 
-##### Design mockup features
+##### 设计原型的特征
 
-Since we need to insert one screenshot for each word, this becomes a bottleneck when training the network ([example](https://docs.google.com/spreadsheets/d/1xXwarcQZAHluorveZsACtXRdmNFbwGtN3WMNhcTdEyQ/edit#gid=0)). Instead of using the images, we extract the information we need to generate the markup.
+因为我们需要为每个单词插入一个截屏，这将会成为训练神经网络[案例](https://docs.google.com/spreadsheets/d/1xXwarcQZAHluorveZsACtXRdmNFbwGtN3WMNhcTdEyQ/edit#gid=0))的瓶颈。因此我们抽取生成标记语言所需要的信息来替代直接使用图像。
```
Comment (sunshine940326) [inline TODO/turning-design-mockups-into-code-with-deep-learning-1.md]: )的瓶颈 -> 多个一个左括号

File: TODO/turning-design-mockups-into-code-with-deep-learning-1.md [hunk 12]
```diff
@@ -178,147 +177,149 @@ Here we make the prediction:
...
+对于另一个平行的过程，其中图像特征首先会展开成一个向量，然后再馈送到一个全连接层而抽取出高级特征。这些图像特征随后会与标记特征相级联而作为编码器的输出。
 
-This can be hard to wrap your mind around - so let’s break it down.
+这个有点难理解，让我来分步描述一下。
 
-##### Markup features
+##### 标记特征
 
-Here we run the word embeddings through the LSTM layer. In this graphic, all the sentences are padded to reach the maximum size of three tokens.
+如下图所示，现在我们将词嵌入投入到 LSTM 层中，所有的语句都填充上最大的三个记号。。
```
Comment (sunshine940326) [inline TODO/turning-design-mockups-into-code-with-deep-learning-1.md]: 所有的语句都填充上最大的三个记号。。 -> 多了一个句号

**Comment by sunshine940326** [comment]
@sakila1012 翻译的很棒，提出 一些小问题，校队完毕

**Comment by sakila1012** [comment]
@sunshine940326 非常感谢，校对的非常认真，细致。

**Comment by sakila1012** [comment]
@leviding 修改完毕

**Comment by leviding** [comment]
@sakila1012 已经 merge 啦~ 快快麻溜发布到[掘金](https://juejin.im)然后给我发下链接，方便及时添加积分哟。

掘金翻译计划有自己的知乎专栏，你也可以投稿哈，推荐使用一个[好用的插件](https://chrome.google.com/webstore/detail/minimalist-markdown-edito/pghodfjepegmciihfhdipmimghiakcjf?hl=zh-CN)。
专栏地址：https://zhuanlan.zhihu.com/juejinfanyi

**Comment by sakila1012** [comment]
@leviding https://juejin.im/post/5a72744e6fb9a01cb64f1d66
